update modular_modernbert -- add inputs_embeds param to ModernBertModel

jxmorris12 opened the pull request:
# What does this PR do?
Hi! Congrats on the release of ModernBERT; it looks amazing. I'm interested in using ModernBERT eventually to train a new [Contextual Document Embeddings](https://arxiv.org/abs/2410.02525) model. 

One desired feature is to pass the contextual and word embeddings together in the second stage, which requires setting the `inputs_embeds` kwarg so that we can pass hidden states directly. This is a feature of typical BERT and other transformer implementations but isn't yet allowed by ModernBERT, so I added it. It's only a few additional lines of code.

cc: @warner-benjamin @tomaarsen @orionw @staghado @bclavie @NohTow @ArthurZucker

jxmorris12 committed:
update modular_modernbert -- add inputs_embeds param to ModernBertModel

jxmorris12 committed:
Merge branch 'main' into patch-1

ArthurZucker reviewed (commented):
Thanks! Yep would be nice to add. let's make it a bit simpler (si how llama has less if elses) and add a small test ! 🤗

tomaarsen committed:
Fix implementation issues; extend to other classes; docstring

First of all, the inputs_embeds shouldn't fully replace `self.embeddings(input_ids)`, because this call also does layer normalization and dropout. So, now both input_ids and inputs_embeds is passed to the ModernBertEmbeddings, much like how BertEmbeddings is implemented.

I also added `inputs_embeds` to the docstring, and propagated the changes to the other model classes.

I also introduced an error if input_ids and input_embeds are both or neither provided.

Lastly, I fixed an issue with device being based solely on input_ids with attention_mask.

tomaarsen committed:
Propagate inputs_embeds to ModernBertForMaskedLM correctly

Also reintroduce inputs_embeds test

tomaarsen commented:
Hello @jxmorris12, @ArthurZucker,

I pushed some changes into this PR to get it closer to completion. Let me know if you're not okay with this, and you can easily revert or delete the commits.

The changes:
1. The `inputs_embeds` shouldn't fully replace `self.embeddings(input_ids)`, because this call also does layer normalization and dropout. So, now both `input_ids` and `inputs_embeds` is passed to the `ModernBertEmbeddings`, much like how `BertEmbeddings` is implemented.
2. I also added `inputs_embeds` to the docstring, and propagated the changes to the other model classes.
3. I also introduced an error if `input_ids` and `input_embeds` are both or neither provided.
4. I fixed an issue with device being based solely on `input_ids` with `attention_mask`.
5. Fix a test, and reintroduce another test.

> let's make it a bit simpler (si how llama has less if elses)

This is sadly not as simple as it seems due to `_unpad_modernbert_input`.

- Tom Aarsen

NohTow commented:
Hello,

Thanks for opening this PR, I really love your work on CDE and I look forward to seeing how it will perform with ModernBERT!
FYI, this feature (being able to pass input_embeds) has also been asked for other use cases [here](https://github.com/AnswerDotAI/ModernBERT/issues/165), so besides retro-compatibility and CDE, it seems like this feature is used by the community and is thus a very cool addition.

I'll make a more thorough review when I come back from vacation this week, but I already checked how to implement it for the linked issue and it seems to be in line with the latest change from Tom.

HuggingFaceDocBuilderDev commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/transformers/pr_35373). All of your documentation changes will be reflected on that endpoint. The docs are available until 30 days after the last update.

jxmorris12 commented:
> Hello @jxmorris12, @ArthurZucker,
> 
> I pushed some changes into this PR to get it closer to completion. Let me know if you're not okay with this, and you can easily revert or delete the commits.
...

thanks Tom! Looks great.

jxmorris12 commented:
Hi @NohTow @ArthurZucker just following up on this -- thanks :)

znsoftm commented:
when will it be merged into the main branch?

ArthurZucker commented on src/transformers/models/modernbert/modular_modernbert.py:
```diff
@@ -1017,35 +1027,49 @@ def forward(
 
         if batch_size is None and seq_len is None:
-            batch_size, seq_len = input_ids.shape[:2]
+            if inputs_embeds is not None:
+                batch_size, seq_len = inputs_embeds.shape[:2]
+            else:
+                batch_size, seq_len = input_ids.shape[:2]
```
if you first embed input ids you don't need two branches!

ArthurZucker commented on src/transformers/models/modernbert/modular_modernbert.py:
```diff
@@ -1017,35 +1027,49 @@ def forward(
-                    input_ids, indices, cu_seqlens, max_seqlen, *_ = _unpad_modernbert_input(
-                        inputs=input_ids, attention_mask=attention_mask
+                if inputs_embeds is None:
+                    with torch.no_grad():
+                        input_ids, indices, cu_seqlens, max_seqlen, *_ = _unpad_modernbert_input(
+                            inputs=input_ids, attention_mask=attention_mask
+                        )
+                else:
+                    inputs_embeds, indices, cu_seqlens, max_seqlen, *_ = _unpad_modernbert_input(
+                        inputs=inputs_embeds, attention_mask=attention_mask
                     )
```
only the input is different here! if you first embed the input, then you don't need two branches!

jxmorris12 replied:
Well the old code contains a torch.no_grad here. That's why there are two branches. Did you all mean to wrap the unpad function in a no_grad? If so, we need to keep it. We definitely can't unpad inputs_embeds in a no_grad block because it will break gradient flow

tomaarsen replied:
I believe #35386 is a more detailed bug report on the gradient breaks

ArthurZucker replied:
My bad! Indeed for input embedding you need grads if you are training an encoder!

ArthurZucker commented on src/transformers/models/modernbert/modular_modernbert.py:
```diff
@@ -1201,19 +1226,32 @@ def forward(
 
         if self.config._attn_implementation == "flash_attention_2":
             if indices is None and cu_seqlens is None and max_seqlen is None:
```
TBH it might make more sense! This way only the base model unpads, and the other models can freely do the unpadding. Kinda up to you 🤗

ArthurZucker reviewed (commented):
Thanks for updating, let's remove complexity and good to go!

NohTow commented:
Hello,
Sorry for the delay.
As mentioned earlier, the code looks good to me.
I only have two small nitpick. As mentioned by @ArthurZucker, first embedding the input_ids should be cleaner and less error prone.
The second is that the XOR test is somewhat a bit less informative than checking if the two are not set together and then checking if at least one is defined. Maybe it's just me, but I find `"You must specify exactly one of input_ids or inputs_embeds"` a bit ambiguous, yet I don't have a better wording.  

Besides those very small details, everything looks good!

NohTow commented:
@jxmorris12 I am very sorry my message was unclear. I did not mean that the code was incorrect nor that testing the two conditions sequentially was cleaner, I just meant that from my reading, the error message could be ambiguous (in the case where the user do not feed anything, the _exactly_ might not be explicit enough).
But again, that is really a nitpick and probably comes from me not properly reading the message, let's ignore that!

jxmorris12 commented:
@jon-tow Oh okay, that makes sense to me! I agree it could be a really confusing error for the user. But maybe since it's in a lot of files (such as LLAMA) we could open a separate issue to improve that error message everywhere?

NohTow commented:
@jxmorris12 Yeah totally, I raised that because I compared the implementations from BERT and this one and the BERT one does the check sequentially and thus have more informative messages, but since Arthur raised that it is already done like that in other models and is standard now, let's just follow the standard!

jxmorris12 committed:
Merge branch 'main' into patch-1

ArthurZucker reviewed (commented):
My comments are not adressed 😓 
- let's embed inputs, then we computes shapes
- let's never require grad on unpadding / padding as it's weight agnostic

NohTow commented:
For the latter, I think it is related to [this](https://github.com/huggingface/transformers/pull/35404).
It was discussed during the original PR, but it breaks the gradients.
I'll let @warner-benjamin give more information.

ArthurZucker commented:
Got it! Okay, for output, we do need the gradients, my bad! 
TLDR: 
- when training: with torch no grad for input, but you need output gradients
- when training: with input embeddings: torch grad because you need propagation 
- when inference, never grad, but usually people do this outside the modeling. 

So the only "gain" is using no grad on input ids all the time

ArthurZucker reviewed (approved):
Thanks for answering my questions! Let's go 🤗
